### ISSUE TITLE
catalog model: support additional System relations

### ISSUE THREAD

**Issue by brethubbard** [backstage/backstage#7109]
## Feature Suggestion

I would like support for additional relations on the System entity.

## Possible Implementation

On the System entity there could be a `spec.consumesApis` or `spec.dependsOn` relation defined explicitly.

Alternatively, a visibility could be declared on an API entity and then we wouldn't have to declare the system then we could derive the public APIs from the sub-components that are part of the system. This would mean that all API relations would be inferred through a component.

## Context

The docs say

> With increasing complexity in software, systems form an important abstraction level to help us reason about software ecosystems.

Part of understanding those systems is understanding its boundaries. That is covered in part by an API that is part of a system but a critical part of that is surfacing the APIs that system depends on.

**Comment by Fox32** [comment]
That's a discussion we already had before. But I'm having trouble finding it... Maybe I find it again 😁

There were different discussion, either having it as relations, or having a flag or level on the API that marks their level of being "public".

**Comment by brethubbard** [comment]
Oh that's interesting, I spent some time trying to determine if there was guidance on this. Anything that I could find that was related was different enough that I thought a new issue made the most sense. If you can find that conversation I would love to take a look at it!

**Comment by Fox32** [comment]
[This](https://github.com/backstage/backstage/pull/3957) could be an entrypoint into researching the discussion.

**Comment by freben** [comment]
Hi!

Yes, interesting and complex subject! :) As you may have seen, it was always intended that systems should serve as this boundary and that it should clarify which of its constituent APIs were intended for outside consumption. It would also give system owners a clearer overview of the collected dependencies of its constituent parts.

Now it sounds like you are suggesting movement in that direction, which is great. We've been working hard in other areas for a while and not leaving a whole lot of room for thinking about this, but let's ponder it and see where it goes.

There are at least two parts to it: what the input should be, and what the output should be. On the output side, having `providesApi`/`apiProvidedBy` relations between the system and some of its APIs might make a lot of sense. So what would drive the generation of those relations? You suggest a `visibility` declaration on the API side, and I think that's a reasonable approach. It puts the responsibility and power clearly in the hands of the API owner, which may not have the ability to change the system itself as easily. I'd consider naming it `scope` perhaps, with values such as `private` (default), `internal`, and `external` (directly accessed by external clients, e.g. the general public) or something like that. Naming is hard!

In practical terms, I think that also is fairly simple to achieve in code. The [builtin kinds processor](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.ts) could emit one more pair of relations on top of the already existing `partOf`/`hasPart`, when the scope level is high enough and a system is given.

Partial overlap: https://github.com/backstage/backstage/issues/4616. That one talks about the "visual" consumption patterns in the user interface, rather than technical consumption patterns. This also highlights the importance of clearly naming of these things - and how there's a risk of shifting purposes and "lending" values over time if their definitions aren't crisp and tight :)

**Comment by jcarres-mdsol** [comment]
I'm going to leave here a comment carried over from https://github.com/backstage/backstage/issues/9014

To me the most natural behavior for system is to roll up the information from the components in the system. 

According to [Backstage docs](https://backstage.io/docs/features/software-catalog/system-model#system), `A system, in this sense, is a collection of resources and components that exposes one or several public APIs`. And unless we want to re-declare APIs, those are going to be defined already in lower level components. So to keep single source of truth we should be able to expose them. Additionally we may want to consider the behavior of `providesApi/apiProvidedBy` does it override or does it augment, etc. But I think an MVP is the rollup behavior, then further extensions can be designed and implemented later as use-cases emerge.

**Comment by freben** [comment]
I agree. Rollup is probably sensible, both from an end user perspective (it's easy to reason about and update the catalog info files to get the desired effect) and from a technical perspective (you can emit the relations easily directly from the processing of the component/api entity).

**Comment by unixunion** [comment]
Is there any progress on this rollup / visibility of internal / external system API's? As we have complex systems and would like to catalog our API's both public and internal to systems.

**Comment by freben** [comment]
Not that I've heard. I still think the concept makes sense.

**Comment by freben** [comment]
You can still make a custom extension to the model for this to be clear, whether it's in the core one or not. For example, you could add a custom field `spec.scope: system` or something along those lines, and then a custom processor that translates that to providesApi/apiProvidedBy relation pairs with the system that it's part of.

**Comment by unixunion** [comment]
@freben the rollups would be very interesting as a feature, as Software Architects, we often need to view several domains or systems, but this gets very noisy when doing so with backstage because of all the domains, systems, components, api and resources all competing for render space on the graph. 

If you turn off specific kinds, and relationships,  you lose the full picture. It would be awesome to have the ability to collapse systems and or domains, and then see the calls as being between the systems or domains instead of between API's and components. 

Much like how https://structurizr.com/ implements it. If we had the rollups / collapse feature, it becomes much easier to use the graph to reason about the system.

**Comment by martina-if** [comment]
Bump! I'd love to see this rollup feature as well. It can make the graphs even more useful because you can explore the architecture diagram in "layers" going from high level to deeper details. We have a couple of adopters asking for this 😁

**Comment by jhaals** [comment]
We _believe_ it will but it's probably worth a try 😅

**Comment by dweber019** [comment]
Just released the following plugin, which could support some use cases here
https://github.com/dweber019/backstage-plugins/tree/main/plugins/relations

Feedback welcome 😉
